### PR TITLE
encapsulate noise formatting

### PR DESCRIPTION
**Pull request by erichulburd**
* replace `format *compiler-noise*` with `format-noise` which includes check to see if defined.
* `format-noise` includes fresh line.

### CODE REVIEW

File: src/addresser/embed-swap.lisp [hunk 1]
```diff
@@ -89,29 +89,29 @@ instruction if it exists, and errors otherwise."
...
-            ;; TODO: this assumes only SWAPs exist in the permutation list
-            (when (or (null best-cost-so-far) ; we have to make progress.
-                      (cost-< new-cost best-cost-so-far))
-              (setf best-cost-so-far new-cost)
-              (setf link-index index))))))
+		       ;; TODO: this assumes only SWAPs exist in the permutation list
+		       (when (or (null best-cost-so-far) ; we have to make progress.
+				 (cost-< new-cost best-cost-so-far))
+			 (setf best-cost-so-far new-cost)
+			 (setf link-index index))))))
```
Comment (appleby) [inline src/addresser/embed-swap.lisp]: hard to tell from diff if this fixed or introduced bad indentation

File: src/addresser/embed-swap.lisp [hunk 2]
```diff
@@ -167,17 +167,15 @@ rewiring REWIRING to the TARGET-REWIRING."
...
       (update-rewiring working-l2p q0 q1)
-      (format *compiler-noise-stream*
-              "EMBED-SWAP: This is a free swap. :)~%~
-               EMBED-SWAP: New rewiring: ~a~%~
-               EMBED-SWAP: New initial rewiring: ~a~%"
-              working-l2p initial-l2p))
+      (format-noise
+              "EMBED-SWAP: This is a free swap. :)~%~ EMBED-SWAP: New
+               rewiring: ~a~%~ EMBED-SWAP: New initial rewiring: ~a"
+               working-l2p initial-l2p))
```
Comment (appleby) [inline src/addresser/embed-swap.lisp]: indentation is wrong and also the re-flowed format string is not equivalent. The trailing `~`s in the old version of the format string mean to ignore the newline and following whitespace, whereas `~ ` is not a valid control char, so `format` will probably complain

http://www.lispworks.com/documentation/HyperSpec/Body/22_cic.htm

File: src/addresser/embed-swap.lisp [hunk 3]
```diff
@@ -167,17 +167,15 @@ rewiring REWIRING to the TARGET-REWIRING."
...
      (t
       ;; in this case, this swap has to be performed by the QPU.
       ;; apply the link swap to working-l2p
       (update-rewiring working-l2p q0 q1)
-      (format *compiler-noise-stream*
-              "EMBED-SWAP: New rewiring: ~a~%"
-              working-l2p)
+      (format-noise
+              "EMBED-SWAP: New rewiring: ~a" working-l2p)
```
Comment (appleby) [inline src/addresser/embed-swap.lisp]: nix newline

File: src/addresser/temporal-addresser.lisp [hunk 4]
```diff
@@ -374,10 +374,9 @@ following swaps."
              instr)
      (when dry-run-escape
        (funcall dry-run-escape))
-     (format *compiler-noise-stream*
-             "DEQUEUE-GATE-APPLICATION: ~/quil:instruction-fmt/ is a 1Q ~
-instruction, adding to logical queue.~%"
-             instr)
+     (format-noise
+             "DEQUEUE-GATE-APPLICATION: ~/quil:instruction-fmt/ is a
+1Q ~ instruction, adding to logical queue."  instr)
```
Comment (appleby) [inline src/addresser/temporal-addresser.lisp]: bad indentation & bad reflow (tilde-newline)

File: src/compilation-methods.lisp [hunk 5]
```diff
@@ -63,13 +63,14 @@
                    (handler-case
                        (let ((result (funcall compilation-method instruction :context context)))
                          (let ((*print-pretty* nil))
-                           (format *compiler-noise-stream*
-                                   "APPLY-TRANSLATION-COMPILERS: Applying ~A to ~/quil:instruction-fmt/.~%"
+                           (format-noise
+                                   "APPLY-TRANSLATION-COMPILERS: Applying ~A to ~/quil:instruction-fmt/."
```
Comment (appleby) [inline src/compilation-methods.lisp]: indent looks off

File: src/compilation-methods.lisp [hunk 6]
```diff
@@ -84,8 +85,8 @@
       ;; if those fail, try the global compilers
       (map nil #'try-compiler (chip-specification-generic-compilers chip-spec))
       ;; if those failed too, there's really nothing more to do.
-      (format *compiler-noise-stream*
-              "APPLY-TRANSLATION-COMPILERS: Could not find a compiler for ~/quil:instruction-fmt/.~%"
+      (format-noise
+              "APPLY-TRANSLATION-COMPILERS: Could not find a compiler for ~/quil:instruction-fmt/."
```
Comment (appleby) [inline src/compilation-methods.lisp]: indent

File: src/compilation-methods.lisp [hunk 7]
```diff
@@ -241,8 +242,8 @@ Returns a value list: (processed-program, of type parsed-program
                    (redirect-edge blk fresh-block (outgoing registrant)))
              ;; try again, but with this new jump
              (push (list fresh-block registrant) block-stack)
-             (format *compiler-noise-stream* "COMPILER-HOOK: Introduced ~A to deal with the rewiring.~%"
-                     (basic-block-name fresh-block))))
+             (format-noise "COMPILER-HOOK: Introduced ~A to deal with the rewiring."
+			   (basic-block-name fresh-block))))
```
Comment (appleby) [inline src/compilation-methods.lisp]: indent

File: src/compilation-methods.lisp [hunk 8]
```diff
@@ -325,11 +326,11 @@ Returns a value list: (processed-program, of type parsed-program
...
-               (format *compiler-noise-stream* "COMPILER-HOOK: Matching rewiring from ~A (~A) to ~A (~A).~%"
-                       (basic-block-name registrant)
-                       final-l2p
-                       (basic-block-name blk)
-                       initial-l2p)
+               (format-noise "COMPILER-HOOK: Matching rewiring from ~A (~A) to ~A (~A)."
+			     (basic-block-name registrant)
+			     final-l2p
+			     (basic-block-name blk)
+			     initial-l2p)
```
Comment (appleby) [inline src/compilation-methods.lisp]: indent

File: src/compilation-methods.lisp [hunk 9]
```diff
@@ -341,7 +342,7 @@ Returns a value list: (processed-program, of type parsed-program
              (return-from process-block))
 
            (let ((*print-pretty* nil))
-             (format *compiler-noise-stream* "COMPILER-HOOK: Visiting ~A for the first time, coming from ~A (~A).~%"
+             (format-noise "COMPILER-HOOK: Visiting ~A for the first time, coming from ~A (~A)."
```
Comment (appleby) [inline src/compilation-methods.lisp]: indent

File: src/compilation-methods.lisp [hunk 10]
```diff
@@ -351,12 +352,12 @@ Returns a value list: (processed-program, of type parsed-program
...
-              (push (list false-target blk) block-stack))
-             ((unconditional-edge target)
-              (push (list target blk) block-stack))
-             (terminating-edge nil))
+		      ((conditional-edge _ true-target false-target)
+		       (push (list true-target blk) block-stack)
+		       (push (list false-target blk) block-stack))
+		      ((unconditional-edge target)
+		       (push (list target blk) block-stack))
+		      (terminating-edge nil))
```
Comment (appleby) [inline src/compilation-methods.lisp]: revert

File: src/compilers/approx.lisp [hunk 11]
```diff
@@ -709,8 +709,8 @@ NOTE: This routine degenerates to an optimal 2Q compiler when *ENABLE-APPROXIMAT
       (dolist (circuit-crafter crafters)
         (unless (and (first candidate-pairs)
                      (double= 1d0 (car (first candidate-pairs))))
-          (format *compiler-noise-stream*
-                  "~&APPROXIMATE-2Q-COMPILER: Trying ~A on ~A...~%"
+          (format-noise
+                  "~&APPROXIMATE-2Q-COMPILER: Trying ~A on ~A..."
```
Comment (appleby) [inline src/compilers/approx.lisp]: indent

File: src/compressor/compressor.lisp [hunk 12]
```diff
@@ -264,46 +264,46 @@ other's."
...
-                       ;; any noncommuting instructions past each other, we
-                       ;; pick a 'bottleneck' instruction.
-                       (let ((insertion-point (find-safe-insertion-node relevant-nodes-for-inspection))
-                             (new-node (rewind-node (first nodes-for-inspection) *rewriting-peephole-size*)))
-                         (splice-instrs-in-at-node output insertion-point)
-                         (mapc #'delete-node relevant-nodes-for-inspection)
-                         (return-from apply-rules new-node))))
-                 (compiler-does-not-apply () nil))))))
+		   ;; make sure we have enough terms, then apply the rule's consumer
+		   (with-simple-restart (try-next-compiler "Ignore this error and try the next rewrite rule in the list.")
```
Comment (appleby) [inline src/compressor/compressor.lisp]: the previous indentation looks correct here

File: src/compressor/compressor.lisp [hunk 13]
```diff
@@ -328,21 +328,21 @@ other's."
...
-                           ((not exhausted-specializations)
-                            (setf exhausted-specializations t)
-                            (chip-specification-generic-rewriting-rules chip-specification))
-                           ;; otherwise, give up.
-                           (t nil))))
-                     (let ((node-to-jump-to (apply-rules rewrite-rules nodes-for-inspection)))
-                       (when node-to-jump-to
-                         (setf node node-to-jump-to)
-                         (go :ADVANCE-NODE))))))))
+                    ((rewrite-rules
```
Comment (appleby) [inline src/compressor/compressor.lisp]: previous indentation was correct

File: src/compressor/compressor.lisp [hunk 14]
```diff
@@ -581,54 +581,58 @@ other's."
...
+        result-instructions))))
 
 
 (defun compress-instructions-with-possibly-unknown-params (instructions chip-specification context &optional processed-instructions)
   "Dispatch routine for compressing a sequence of INSTRUCTIONS, perhaps with unknown parameter values sprinkled through, based on the routines specified by a CHIP-SPECIFICATION and the current CONTEXT."
-  (format-quil-sequence *compiler-noise-stream*
-                        instructions
-                        "COMPRESS-INSTRUCTIONS: Selected the following sequence for compression:~%")
+  (when *compiler-noise*
+    (format-quil-sequence *compiler-noise-stream*
```
Comment (appleby) [inline src/compressor/compressor.lisp]: `*compiler-noise-stream*` -> `*compiler-noise*`

File: src/compressor/compressor.lisp [hunk 15]
```diff
@@ -783,233 +787,233 @@ This specific routine is the start of a giant dispatch mechanism. Its role is to
...
-                              (dolist (subaddress (coerce (vnth suborder (hardware-object-cxns obj)) 'list))
-                                (let ((subgovernor (nth subaddress (nth suborder governors))))
-                                  (assert (not (typep (first (governed-queue-contents subgovernor)) 'application)))
-                                  (set-gq-fields subgovernor ':empty nil (make-null-resource))))))))
-                       ;; now set us up as fragile and start dumping instructions in
-                       (set-gq-fields governed-queue ':fragile nil (make-null-resource))
-                       (dolist (instr compressed-instructions)
-                         (process-instruction instr))
-                       (set-gq-fields governed-queue ':empty nil (make-null-resource)))))))
+			    ;;
```
Comment (appleby) [inline src/compressor/compressor.lisp]: revert

File: src/addresser/addresser-common.lisp [hunk 16]
```diff
@@ -556,24 +556,24 @@ If DRY-RUN, this returns T as soon as it finds an instruction it can handle."
...
+      (format-noise
+       "DEQUEUE-BEST-INSTR: Elected to schedule ~/quil:instruction-fmt/."
+       instr)
       (let ((rewired-instr (copy-instance instr)))
         (rewire-l2p-instruction working-l2p rewired-instr)
-        (format *compiler-noise-stream*
-                "DEQUEUE-BEST-INSTR: ~/quil:instruction-fmt/ is ~/quil:instruction-fmt/ in the current rewiring~%"
-                instr rewired-instr)
+        (format-noise
+         "DEQUEUE-BEST-INSTR: ~/quil:instruction-fmt/ is ~/quil:instruction-fmt/ in the current rewiring"
```
Comment (appleby) [inline src/addresser/addresser-common.lisp]:
```suggestion
         "DEQUEUE-BEST-INSTR: ~/quil:instruction-fmt/ is ~/quil:instruction-fmt/ in the current rewiring."
```

not your fault, I know, but @kilimanjaro will be pleased

File: src/compressor/compressor.lisp [hunk 17]
```diff
@@ -618,17 +617,19 @@ other's."
                    reduced-decompiled-instructions)
                   (t
                    reduced-instructions))))
-          (format-quil-sequence *compiler-noise-stream*
-                                result-instructions
-                                "COMPRESS-INSTRUCTIONS: Replacing the above sequence with the following:~%")
+          (when *compiler-noise*
+            (format-quil-sequence *compiler-noise-stream*
```
Comment (appleby) [inline src/compressor/compressor.lisp]: `*compiler-noise-stream*` -> `*compiler-noise*`

File: src/options.lisp [hunk 18]
```diff
@@ -17,8 +17,16 @@
 (defvar *resolve-include-pathname* 'identity
   "The function used to handle/resolve pathnames passed to INCLUDE directives in Quil. Specifically, it should be a function designator that takes one argument (a pathname designator) and returns the absolute pathname to include, or signals an error.")
 
-(defvar *compiler-noise-stream* (make-broadcast-stream)
-  "The stream used to emit compiler debug output to.  By default, this variable is bound to a stream that suppresses display.")
+(defvar *compiler-noise* nil
+  "The stream used to emit compiler debug output to. This variable can be bound to a stream that suppresses display, eg (make-broadcast-stream).")
```
Comment (appleby) [inline src/options.lisp]:
```suggestion
  "The stream on which to emit compiler debug output or NIL to suppress compiler noise.")
```

File: src/options.lisp [hunk 19]
```diff
@@ -17,8 +17,16 @@
...
+
+(defmacro format-noise (str &rest args)
+  "format-noise checks to see whether *compiler-noise* has been set and, if so, formats it according to the passed format string and arguments."
```
Comment (appleby) [inline src/options.lisp]:
```suggestion
  "FORMAT-NOISE checks to see whether *COMPILER-NOISE* has been set and, if so, formats it according to the passed format string and arguments."
```

We (usually) follow the convention of writing symbol names in all caps in docstrings.

File: app/src/entry-point.lisp [hunk 20]
```diff
@@ -481,16 +481,14 @@
          (setf *human-readable-stream* *error-output*)
          (setf *quil-stream* *standard-output*)
 
-         (when verbose
-           (setf *verbose* *human-readable-stream*))
```
[the remark this answers is not in the document]
Comment (appleby) [reply app/src/entry-point.lisp]: ^ still think we should remove `*verbose*` entirely if it's no longer used.

File: src/compilation-methods.lisp [hunk 21]
```diff
@@ -63,9 +63,9 @@
...
                        (let ((result (funcall compilation-method instruction :context context)))
                          (let ((*print-pretty* nil))
-                           (format *compiler-noise-stream*
-                                   "APPLY-TRANSLATION-COMPILERS: Applying ~A to ~/quil:instruction-fmt/.~%"
-                                   compilation-method instruction))
+                           (format-noise
+                            "APPLY-TRANSLATION-COMPILERS: Applying ~A to ~/quil:instruction-fmt/."
+                            compilation-method instruction))
                          (dolist (instr result)
                            (write-string "    " *compiler-noise-stream*)
```
Comment (appleby) [inline src/compilation-methods.lisp]: missed a spot

File: src/compressor/compressor.lisp [hunk 22]
```diff
@@ -607,7 +605,8 @@ other's."
                                                        reduced-instructions
                                                        reduced-decompiled-instructions
                                                        context)
-    
+
+
```
Comment (appleby) [inline src/compressor/compressor.lisp]: will let this one slide 😀

File: src/compilation-methods.lisp (hunk 21 above)
Comment (appleby) [reply src/compilation-methods.lisp]: ditto just below

**Comment by appleby** [comment]
Oh, and don't forget to take stylewarning's suggestion [here](https://github.com/rigetti/quilc/issues/535#issuecomment-571295128) and time the test suite before and after so you can brag about those sweet sweet speedups (or cry if it didn't do anything :D).

For example, in slime repl

```
(ql:quickload :cl-quil-tests)
(time (cl-quil-tests:run-cl-quil-tests))
```

File: app/src/entry-point.lisp (hunk 20 above)
[the remark this answers is not in the document]
Comment (appleby) [reply app/src/entry-point.lisp]: still need to remove the `(defparameter *verbose* ...)` from `app/src/globals.lisp`

[the remark this answers is not in the document]
Comment (appleby) [reply app/src/entry-point.lisp]: just waiting on this last change before doing a final test run + approve

**Comment by erichulburd** [comment]
Context: 2919 Lambda Quad: `Linux lambda-quad 4.15.0-72-generic #81-Ubuntu SMP Tue Nov 26 12:20:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux`

`SBCL 1.5.8.27-b8ee8710a`

## erichulburd:enhancement/compiler_noise_macro_ii
(#<test-run of CL-QUIL-TESTS: 212 tests, 40098 assertions, 0 failures in 541.085 sec>)

## master
(#<test-run of CL-QUIL-TESTS: 212 tests, 40098 assertions, 0 failures in 589.467 sec>)

**Comment by appleby** [comment]
40 sec, not too shabby!

**Comment by appleby** [comment]
> 40 sec, not too shabby!

unless the stddev is +/- 60s :)

File: src/compilation-methods.lisp [hunk 23]
```diff
@@ -63,13 +63,14 @@
...
-                           (print-instruction instr *compiler-noise-stream*)
-                           (terpri *compiler-noise-stream*))
+                           (format-noise
+                            "APPLY-TRANSLATION-COMPILERS: Applying ~A to ~/quil:instruction-fmt/."
+                            compilation-method instruction))
+                         (when *compiler-noise*
+                           (dolist (instr result)
+                             (write-string "    " *compiler-noise*)
+                             (print-instruction instr *compiler-noise*)
+                             (terpri *compiler-noise*)))
```
Comment (appleby) [inline src/compilation-methods.lisp]: mentioned in slack, but leaving a comment here so I don't forget over the weekend...

I think it makes sense to replace this whole `when`/`dolist` with a call to `format-noise` here, something like:

```lisp
(format-noise "~{    ~/quil:instruction-fmt/~%~}" result)
```

this cuts down on the number of explicit references to `*compiler-noise*`, plus we have been slowly converting `print-instruction` calls to `format` + `~/quil:instruction-fmt/` anyway

**Pull request merged**